### PR TITLE
Ir updates (fixed)

### PR DESCRIPTION
**Pull request by 0rphon**
## Motivation

This is a second PR for the #414 PR that had to be reverted because i failed to update its cargo.toml versions before merging.

This adds support for new leo features in the IR. Another branch currently exists on leo that implements IR using these changes. 

Adds:
- support for reverse/inclusive ranges
- bug fixes for complex array indexing
- support for unsized arrays
- moves the len method from leo into the eval core

## Test Plan

all tests are passing. once this branch and the leo branch are merged for DP3 it will be tested against the fuzzer

## Related PRs

follow up to #218
developed alongside the [leo ir branch](https://github.com/AleoHQ/leo/tree/ir)

### CODE REVIEW

**Review by Protryon — APPROVED** [review]

File: marlin/src/marlin/errors.rs [hunk 1]
```diff
@@ -61,6 +61,15 @@ impl From<snarkvm_polycommit::Error> for MarlinError {
     }
 }
 
+impl MarlinError {
+    pub fn into_snark_error(&self, prefix: &str) -> SNARKError {
```
Comment (collinc97) [inline marlin/src/marlin/errors.rs:65]: Is there a reason why we implemented a standalone method to convert this error?
The `from` method is implemented for `SNARKError` just below.

File: marlin/src/parameters.rs [hunk 2]
```diff
@@ -61,7 +61,7 @@ where
     /// Creates an instance of `Parameters` from a given universal SRS.
     pub fn new<C: ConstraintSynthesizer<E::Fr>>(circuit: &C, universal_srs: &SRS<E>) -> Result<Self, SNARKError> {
         let (proving_key, verifying_key) = MarlinTestnet1::<E>::circuit_setup(universal_srs, circuit)
-            .map_err(|error| SNARKError::Crate("marlin", format!("could not index - {:?}", error)))?;
+            .map_err(|error| error.into_snark_error("could not index"))?;
```
Comment (collinc97) [inline marlin/src/parameters.rs:64]: See `into_snark_error()` comment

File: marlin/src/snark.rs [hunk 3]
```diff
@@ -114,7 +114,7 @@ where
         let proving_time = start_timer!(|| "{Marlin}::Proving");
         let proof =
             MarlinTestnet1::<E>::prove_with_terminator(&proving_key.proving_key, input_and_witness, terminator, rng)
-                .map_err(|error| SNARKError::Crate("marlin", format!("Failed to generate proof - {:?}", error)))?;
+                .map_err(|error| error.into_snark_error("Failed to generate proof"))?;
```
Comment (collinc97) [inline marlin/src/snark.rs:117]: See `into_snark_error()` comment

File: marlin/src/snark.rs [hunk 4]
```diff
@@ -126,7 +126,7 @@ where
     ) -> Result<bool, SNARKError> {
         let verification_time = start_timer!(|| "{Marlin}::Verifying");
         let res = MarlinTestnet1::<E>::verify(&verifying_key, &input.to_field_elements()?, &proof)
-            .map_err(|_| SNARKError::Crate("marlin", "Could not verify proof".to_owned()))?;
+            .map_err(|error| error.into_snark_error("Could not verify proof"))?;
```
Comment (collinc97) [inline marlin/src/snark.rs:129]: See `into_snark_error()` comment

**Review by collinc97 — CHANGES_REQUESTED** [review]
Thank you for resubmitting @0rphon!

File: marlin/src/marlin/errors.rs (hunk 1 above)
Comment (0rphon) [reply marlin/src/marlin/errors.rs:65]: all of these changes you pointed out are from commit #364 which already exist's on master but got merged into ir-updates while i was updating from master yesterday. so im not sure

**Review by collinc97 — APPROVED** [review]

**Comment by 0rphon** [comment]
adding another reviewer to this since it's lack of reviews is blocking other branches

**Review by gluax — APPROVED** [review]

File: ir/src/values/mod.rs [hunk 5]
```diff
@@ -41,6 +41,25 @@ pub enum Integer {
     I128(i128),
 }
 
+impl TryFrom<Integer> for u32 {
```
Comment (collinc97) [inline ir/src/values/mod.rs:44]: @0rphon @Protryon who is calling this function?
I do not see it being used in snarkVM currently.

**Comment by collinc97** [comment]
@0rphon @Protryon Can we add a readme for the tests run by `setup_test()` in `eval/tests/test.rs`?
If a future snarkvm change breaks a test run by that function, then it would be useful to have a set of instructions on how to 1. find the leo code and 2. regenerate the ir for the test without having to disrupt you guys.

File: ir/src/values/mod.rs (hunk 5 above)
Comment (0rphon) [reply ir/src/values/mod.rs:44]: line 49
https://github.com/AleoHQ/leo/blob/ir/compiler/src/statement/assign/assignee/array_range_index.rs

**Comment by 0rphon** [comment]
@collinc97 so is this good to merge then?

**Pull request merged**